kernel/msg_q.c

```diff
@@ -120,6 +120,14 @@ int k_msgq_cleanup(struct k_msgq *msgq)
 	return 0;
 }
 
+static inline void msgq_copy(char *dest, const char *src, size_t n)
```

Comment:
Isn't @keith-packard fixing picolibc memcpy() to be a better optimized version?  This variant might be faster in the specific benchmark with very small messages, but a bytewise copy is near-pessimal as the copies get larger (like, literally 4x slower at minimum, and more like 16x on bigcore x86/arm64 devices with wide buses and cache-aware SIMD load/store/move operations!).

Really this seems like a bad trade, and we're doing too much benchmark chasing.  If we really need to muck with this, my vote would be to "fix" the API to require alignment (via a kconfig for non-aligned compatibility or whatever) and code a word-chunked copy.

Reply:
agree with @andyross  let's not touch the libc part and focus on kernel scheduling, you can build the optimized picolibc using ```-DCONFIG_PICOLIBC_SPEED_OPTIMIZATIONS=y  -DCONFIG_PICOLIBC_USE_MODULE=y``` or you can use NEWLIB. I just did a quick test:

baseline with optimized picolibc:  

```
*** Booting Zephyr OS build v3.7.0-4722-g338d595a6a19 ***
**** Thread-Metric Message Processing Test **** Relative Time: 30
Time Period Total:  8109040
 ```
 
with your changes in this PR: 
```
*** Booting Zephyr OS build v3.7.0-4726-g3bea9b73042c ***
**** Thread-Metric Message Processing Test **** Relative Time: 30
Time Period Total:  7362848
 ```
 
Your PR without custom byte copy:
```
*** Booting Zephyr OS build v3.7.0-4727-gc8937c0a4269 ***
**** Thread-Metric Message Processing Test **** Relative Time: 30
Time Period Total:  10127756
 ```

Reply:
I have no problems dropping that change.  Incidentally, I did some additional experimentation to profile the copying performance under different copy sizes.

Copying 1 byte at a time: 7273612
Copying 2 bytes at a time: 9601154
Copying 4 bytes at a time: 11429945
Copying 8 bytes at a time: 11650579

I mention these numbers to show the limits of what could be expected from changing the copying methodology (assuming everything else stays the same).

Reply:
On arm32 the advantage of the 8 byte case is just loop unrolling, I don't think (though apply salt, not my arch) that there's a 64 bit load instruction available on non-FPU devices.